### DISCUSSION TITLE
[v2] Add a plugin as a part of tauri app

### DISCUSSION THREAD

**Discussion by pronebird** [tauri-apps/tauri#9337] [Q&A] [answered]
Hi,

I found that plugins are a decent way to break down the complex app onto smaller parts and I'd like to just define one quickly inside of my tauri app.

After defining and wiring up the plugin, all I see is this in my javascript console:

> Plugin did not define its manifest

How can I tackle this quickly? Maybe there is a call I can make in `PluginApi` to get rid of this nonsense and get going?

**Comment by FabianLars** [comment, marked as answer]
inline plugins are unfortunetely not as nice to use as in v1 due to the permission changes. You will have to modify your build.rs file similar to <https://github.com/tauri-apps/tauri/blob/dev/examples/api/src-tauri/build.rs#L9-L12> (it's for https://github.com/tauri-apps/tauri/blob/dev/examples/api/src-tauri/src/menu_plugin.rs)

**Comment by watsom27** [comment]
Are there any plans to change this? Migrating from a v1 app with tens of "inline plugins", each with tens of command handlers is not turning out to be a friendly experience.

Especially when it seems like we also have to "allow list" each command in our own plugins?

**Reply by FabianLars** [reply]
As far as i am aware, the "modify build.rs" requirement will stay, at least in the initial 2.0 release(s), but the last version added an option to generate default permissions for you so you don't have to allow the commands one by one manually. https://github.com/tauri-apps/tauri/pull/10559

**Reply by watsom27** [reply]
Thank you for the reply.

What is the reasoning around keeping the "modify build.rs" requirement? 

Why do we need to whitelist our own code?

**Reply by FabianLars** [reply]
Because the new plugin system was designed without inline plugins in mind (as far as i know at least, i wasn't involved in that) so we're talking about potentially large changes (or maybe not considering that we'd "just" have to do what i linked behind the scenes). Do not overlook the `at least in the initial 2.0 release(s)` part though, 2.0 is basically in a feature freeze right now and the team is generally really busy right now.

I'm also not sure if we keep the inline plugin concept in v3 etc or switch to something else, considering that most people only started using them because they didn't want to have **one** huge list of commands which is something that could be solved differently. But again, not really involved there 🤷

**Reply by watsom27** [reply]
> considering that most people only started using them because they didn't want to have one huge list of commands

This is likely incorrect, the documentation directs people towards using inline plugins: https://v1.tauri.app/v1/guides/features/plugin

Inline plugins offer a very nice way of organising commands both from the Rust side, and from the Frontend side. For example: `invoke("plugin:help_menu|set_selected",...`. It would be very nice if there was a way of doing this (or something similar) in V2.

**Reply by FabianLars** [reply]
> This is likely incorrect, the documentation directs people towards using inline plugins: [v1.tauri.app/v1/guides/features/plugin](https://v1.tauri.app/v1/guides/features/plugin)

Where does it do that? The linked page is for an external plugin.

Edit: Ahh i see, it's the `They are also a very convenient way to structure your code base!` sentence. Alright, fair enough!

**Reply by watsom27** [reply]
Thats the one, and it's also a very accurate line! They _are_ a very convinient way to structure things, in V1 😅

**Reply by FabianLars** [reply]
Yeah, i myself recommended it many many times so it really boggles my mind how we didn't take that into account when we reworked the plugins for v2.
